classes/profile.php

```diff
@@ -193,12 +193,32 @@ public function save_record(): int {
         $this->raw_set('timemodified', $now);
         $this->raw_set('usermodified', $USER->id);
 
+        $storeduser = (int) $this->raw_get('userid');
+
         // We may not have obtained a valid userid when the profile record was created.
-        // If the current profile userid is 0, but there's a valid $USER->id, update it.
-        if ($USER->id && ! $this->raw_get('userid')) {
+        // If the stored userid is 0, and there's now a valid $USER->id, update the stored userid.
+        if ($USER->id && !$storeduser) {
+            $this->raw_set('userid', $USER->id);
+        }
+
+        // If current userid isn't guest, but stored user is guest, update it.
+        $guestrole = get_guest_role();
+        $guestid = (int) $guestrole->id;
+        if ($USER->id !== $guestid && $storeduser === $guestid) {
             $this->raw_set('userid', $USER->id);
         }
 
+        // If current user isn't cli admin, but stored user is cli admin, update it.
+        if (script_metadata::get_script_type() === self::SCRIPTTYPE_CLI) {
+            $admins = get_admins();
```

Comment:
get_admin() returns the 'primary admin' or main admin. They are the only one which is treated special, the other admins are not a concern. So we don't need the loop over get_admins()

I know this is confusing moodle internals :face_with_spiral_eyes: